### PR TITLE
Requirements for phase 2

### LINKED ISSUE #68: Generics (in delegates, events and methods)(at least two);

### LINKED ISSUE #73: Concurrent programming (threading or async/await (for your own written classes); common resource usage between threads);

### LINKED ISSUE #74: No instances are created using 'new' keyword, dependency injection is used everywhere;

### PR DESCRIPTION
**Pull request by onlyidev**
- Added LockManager for logger (closes #73)
- Added Generic Statistics Processor for lazy statistics (closes #68)
- Added more dependency injection (closes #74)

### CODE REVIEW

**Review by Storas008 — DISMISSED** [review]

File: app/Services/StatisticsProcessor.cs [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    public int GetCount(Selector selector)
+    {
+        if (count.Item1 < 0 || DateTime.Compare(count.Item2.AddSeconds(_interval), DateTime.Now) < 0)
+        {
+            count = Tuple.Create(model.Select(selector.Invoke).Count(), DateTime.Now);
+        }
+        return count.Item1;
+    }
+
+    public async Task<double> getAverageCollAsync(Selector selector, Func<T, double> elems)
```
Comment (Azuolas3) [inline app/Services/StatisticsProcessor.cs:32]: Method names have to be in pascal case

File: app/Services/StatisticsProcessor.cs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        }
+
+        double average = await Task<double>.Run(() => model.Select(selector.Invoke).Average(elems));
+
+        averages.TryAdd(selector, Tuple.Create(average, DateTime.Now));
+
+        return average;
+    }
+
+    public StatisticsProcessor<T> updateDbContext(DbSet<T> newContext)
```
Comment (Azuolas3) [inline app/Services/StatisticsProcessor.cs:47]: Method names have to be in pascal case

File: app/Services/StatisticsProcessor.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+{
+    StatisticsProcessor<T> GetStatisticsProcessor<T>(DbSet<T> modelSet, int interval = 60) where T : class;
+}
+
+public class StatisticsProcessorFactory : IStatisticsProcessorFactory
+{
+
+    private ConcurrentDictionary<Tuple<Type, int>, object> availableProcessors = new ConcurrentDictionary<Tuple<Type, int>, object>();
+
+    private void addProcessor<T>(Tuple<Type, int> key, StatisticsProcessor<T> value) where T : class
```
Comment (Azuolas3) [inline app/Services/StatisticsProcessor.cs:64]: Method names have to be in pascal case

File: app/Services/StatisticsProcessor.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    {
+        availableProcessors.TryAdd(key, value);
+    }
+
+    private StatisticsProcessor<T>? getProcessor<T>(Tuple<Type, int> type) where T : class
```
Comment (Azuolas3) [inline app/Services/StatisticsProcessor.cs:69]: Method names have to be in pascal case

**Review by Azuolas3 — CHANGES_REQUESTED** [review]
Minor changes with code formatting

**Review by Azuolas3 — APPROVED** [review]

**Pull request merged**